// grunt-coffeelint/Gruntfile.js
// Generated by CoffeeScript 1.7.1
(function() {
  module.exports = function(grunt) {
    grunt.initConfig({
      coffeelint: {
        options: {
          indentation: {
            value: 4,
            level: 'warn'
          }
        },
        one: {
          files: {
            src: ['test/fixtures/*.coffee']
          },
          options: {
            indentation: {
              value: 2,
              level: 'warn'
            },
            'no_trailing_semicolons': {
              level: 'warn'
            }
          }
        },
        two: ['test/fixtures/correct.coffee', 'test/fixtures/some.coffee']
      }
    });
    grunt.loadTasks('tasks');
    grunt.loadNpmTasks('grunt-npm');
    grunt.loadNpmTasks('grunt-bump');
    grunt.registerTask('default', 'coffeelint');
    grunt.registerTask('test', 'coffeelint');
    return grunt.registerTask('release', 'Bump version, push to NPM.', function(type) {
      return grunt.task.run(["bump:" + (type || 'patch'), 'npm-publish']);
    });
  };

}).call(this);
